src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs

```diff
@@ -169,20 +169,29 @@ private protected static ObjectNodeSection GetSharedSection(ObjectNodeSection se
                     // For R_ARM_THM_JUMP24 the thumb bit cannot be encoded, so mask it out.
                     long maskThumbBitOut = relocType is IMAGE_REL_BASED_THUMB_BRANCH24 or IMAGE_REL_BASED_THUMB_MOV32_PCREL ? 1 : 0;
                     long maskThumbBitIn = relocType is IMAGE_REL_BASED_THUMB_MOV32_PCREL ? 1 : 0;
+                    long adjustedAddend = addend;
 
-                    addend -= relocType switch
+                    adjustedAddend -= relocType switch
                     {
                         IMAGE_REL_BASED_REL32 => 4,
                         IMAGE_REL_BASED_THUMB_BRANCH24 => 4,
                         IMAGE_REL_BASED_THUMB_MOV32_PCREL => 12,
                         _ => 0
                     };
 
-                    addend += definedSymbol.Value & ~maskThumbBitOut;
-                    addend += Relocation.ReadValue(relocType, (void*)pData);
-                    addend |= definedSymbol.Value & maskThumbBitIn;
-                    addend -= offset;
-                    Relocation.WriteValue(relocType, (void*)pData, addend);
+                    adjustedAddend += definedSymbol.Value & ~maskThumbBitOut;
+                    adjustedAddend += Relocation.ReadValue(relocType, (void*)pData);
+                    adjustedAddend |= definedSymbol.Value & maskThumbBitIn;
+                    adjustedAddend -= offset;
+
+                    if (relocType is IMAGE_REL_BASED_THUMB_BRANCH24 && (uint)adjustedAddend >= 0xFFFFFF)
```

Comment:
On second thought... I am not sure this is the correct check for the signed values.

Reply:
Perhaps I can make the `Relocation.FitsInThumb2BlRel24` public and call it here.